CoMeT: Dynamic Latency and Energy Version with Integrated CACTI Support

sejproo opened the pull request:
This work was done as part of my Master's thesis project. Please refer to the thesis for more details.

siddhulokesh commented:
Hi Chirag,

Thanks for the effort.

Did a quick review. A few questions and suggestions came up:

**1. Regarding the README:**

The current version mentions:

tar -xzvf cacti.tar.gz

Instead of using a .tar.gz archive, would it be better to add CACTI as a Git submodule?

- It ensures version consistency by pinning to a specific CACTI commit.

- Avoids committing large binary files to the repository.

- Makes the setup cleaner and easier for contributors and CI systems.

We could consider the following replacement in the README:

### Initialize CACTI submodule
git submodule update --init --recursive

### Build CACTI
cd cacti
make
cd ..

This would make the setup reproducible and remove the need for manual extraction.

2. Please check if the following **sanity checks pass:**

- make in cacti/ completes without errors

- Latency and energy outputs from CACTI do not contain NaN or unexpected 0.0

- Simulator runs with --enable-cacti on standard benchmarks

- Simulator works without CACTI as well (i.e., backward compatibility is preserved)

3. Do you have an approximate idea of the **simulation slowdown due to dynamic CACTI calls** for DRAM energy and latency.

4. Would it be possible to share a **link to your Master’s thesis** for reference.

If possible, please share your responses by the weekend. Will try to do a more detailed review by then.

Thanks,
Lokesh Siddhu

sejproo commented:
I have modified the code to allow Cacti to be loaded as a submodule linked to the forked Cacti repository, where I fixed and adapted it for CoMeT compatibility. README also adapted accordingly.

sejproo commented:
Here are the answer to your final questions:

**1. CACTI Version:
Which version of CACTI is included in the cacti/ directory? Have any modifications been made to this version? If so, could you briefly (1-2 lines) summarize them?**
The version I used is CACTI 7.0, available at the following link:
https://github.com/HewlettPackard/cacti 
This version eventually includes CACTI-3DD.

The repository I forked and connected to CoMeT as a submodule is here:
https://github.com/sejproo/cacti 

I made two main modifications:
Fixed a bug in parameters.cc, as described in this issue: https://github.com/HewlettPackard/cacti/issues/2 
Modified io.cc to output relevant configuration details (e.g., latency and energy values) in the CSV .out file. This made it easier to implement a lexer and use CACTI variables.


**2. Tracking Modifications:
If modifications were made to CACTI, is there a diff file or any way to identify those changes?**
Yes, all the changes I made are visible in this comparison between my fork and the original repository: https://github.com/HewlettPackard/cacti/compare/master...sejproo:cacti:master


**3. Performance Overhead:
Is there a known runtime overhead when enabling CACTI? (I recall you mentioned it's negligible — just reconfirming whether this holds for large batch simulations.)**
CACTI is executed only once during the configuration setup. If it fails, the simulation falls back to the original implementation without retrying.

The only for and while loops I introduced are related to parsing configuration variables in the CACTI config file. These loops are not performance bottlenecks, as they are based on the number of lines in the config file, most of which are comments, so the total is still negligible.

Even for heavier configurations, e.g., HMCs with 128 layers, CACTI remains fast.


**4. Future Updates:
If CACTI is updated in the future, would it be feasible to reapply your changes, or should we consider freezing this version permanently?**
Yes, it would be feasible to reapply the changes. I only:

Changed one line to fix the bug (as mentioned in point 1),
And a few lines to adjust how the CSV output is generated.

However, note that the lexer I designed assumes a fixed position of the keys vector in simulator.cc (lines 166–200).

Since CACTI hasn't been updated in a long time, perhaps future students can continue working on my forked repository. If desired, I could give you and Anuj access to review pull requests.


**5. CACTI Output Key Limit:
In simulator.cc, the makeConfig function includes a check:
if (keys.size() >= 36)
std::cout << "Error: Too many keys in CACTI output file" << std::endl;
Given that 34 keys are written to the output file afterward, the >= 36 condition seems a bit counter-intuitive. Could you clarify why 36 was chosen and what this "too many keys" error indicates?**

The CSV output currently includes an extra comma after the last variable, which results in a total keys.size() of 35 instead of 34. That’s why the condition checks for >= 36.

<img width="1746" height="1420" alt="image" src="https://github.com/user-attachments/assets/8ca16e75-5a0e-41b2-bf90-4213199751a9" />
(cacti/io.cc)

Now that you’ve pointed it out, I’ll remove this extra comma in the configuration. I also realize it would be safer to check if keys.size() < 34, to avoid potential out-of-bounds access.

Changes have been made accordingly to Cacti and simulator.cc in CoMeT.

---
